pyccel/stdlib/lists/lists.c

```diff
@@ -0,0 +1,665 @@
+
+/**
+ * @brief Set the length and type object for group allocation.
+ * 
+ * @param list 
+ * @param len 
+ * @param type 
+ */
+void set_length_and_type(t_list *list, size_t *len, int8_t *type)
+{
+    t_list *tmp;
+
+    tmp = list;
+    *len = tmp->size;
+    while (tmp->type == lst_list)
+    {
+        tmp = ((t_list **)(tmp->elements))[0];
+        *len *= tmp->size;
+    }
+    *type = tmp->type;
+}
+
+/**
+ * @brief Branch recursively through every element in the list and copy 
+ * it's value in the array pointed to by `int8_t *group`.
+ * Returns the number of elements copied on every recursion.
+ * 
+ * @param list 
+ * @param group 
+ * @param offset 
+ * @return size_t 
+ */
+size_t collect_data(t_list *list, int8_t *group, size_t offset)
+{
+    if (list->type == lst_list)
+    {
+        for (size_t i = 0; i < list->size; i++)
+            offset += collect_data(((t_list **)list->elements)[i], group, offset);
+    }
+    else
+    {
+        memcpy(group + offset*tSizes[list->type], list->elements, list->size);
+        offset += list->size;
+    }
+    return (offset);
+}
+
+/**
+ * @brief Allocate and collect the sub-elements of a list.
+ * 
+ * @param list 
+ * @param len 
+ * @param type 
+ * @return int8_t* 
+ */
+int8_t *group_node_items(t_list *list, size_t len, int8_t type)
+{
+    int8_t *group;
+
+    group = (int8_t *)malloc(len * tSizes[type]);
+    collect_data(list, group, 0);
+    return (group);
+}
+
+/**
+ * @brief If the list consists of pointers over other lists, it attempts
+ * to collect the sub-elements of the lists pointed to by the indexs `i1`
+ * and `i2`, putting them in two seperate allocated arrays, then compare 
+ * every value.
+ * Returns comparision value.
+ * 
+ * @param list 
+ * @param i1 
+ * @param i2 
+ * @return int 
+ */
+int compare(t_list *list, int i1, int i2)
+{
+    int8_t *group_1;
+    int8_t *group_2;
+    int8_t type;
+    size_t len;
+    int cmp;
+
+    cmp = 0;
+    len = 0;
+    type = 0;
+    if (list->type != lst_list)
+    {
+        switch (list->type)
+        {
+            case lst_int8:
+                return *((int8_t *)GET_INDEX(list, i1)) - *((int8_t *)GET_INDEX(list, i2));
+            case lst_int16:
+                return *((int16_t *)GET_INDEX(list, i1)) - *((int16_t *)GET_INDEX(list, i2));
+            case lst_int32:
+                return *((int32_t *)GET_INDEX(list, i1)) - *((int32_t *)GET_INDEX(list, i2));
+            case lst_int64:
+                return *((int64_t *)GET_INDEX(list, i1)) - *((int64_t *)GET_INDEX(list, i2));
+            case lst_float:
+                return *((float *)GET_INDEX(list, i1)) - *((float *)GET_INDEX(list, i2));
+            case lst_double:
+                return *((double *)GET_INDEX(list, i1)) - *((double *)GET_INDEX(list, i2));
+            case lst_complex:
+                fprintf(stderr, "Ordering of complex numbers is not supported.");
+                exit(-1);
+        }
+    }
+
```

Comment:
I would prefer an `else` here. That way the compiler will warn you if a `return` is forgotten in the `if`